File: sip-phone-lib/src/main/java/com/telcobright/sipphone/verto/SdpBuilder.java
package com.telcobright.sipphone.verto;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Builds SDP offers/answers with AMR and PCMU codecs for Verto/FreeSWITCH.
 */
public final class SdpBuilder {

    public static final String CODEC_PCMU = "PCMU";
    public static final String CODEC_AMR_NB = "AMR-NB";
    public static final String CODEC_AMR_WB = "AMR-WB";

    public static final int CODEC_TYPE_NB = 0;
    public static final int CODEC_TYPE_WB = 1;

    private static final int PT_PCMU = 0;
    private static final int PT_AMR_DYNAMIC = 96;

    private SdpBuilder() {}

    public static String buildOffer(String localIp, int rtpPort, String codec) {
        return switch (codec) {
            case CODEC_AMR_NB -> buildAmrOffer(localIp, rtpPort, CODEC_TYPE_NB);
            case CODEC_AMR_WB -> buildAmrOffer(localIp, rtpPort, CODEC_TYPE_WB);
            default -> buildPcmuOffer(localIp, rtpPort);
        };
    }

    private static String buildPcmuOffer(String localIp, int rtpPort) {
        return "v=0\r\n" +
                "o=- " + System.currentTimeMillis() + " 1 IN IP4 " + localIp + "\r\n" +
                "s=SipPhone\r\n" +
                "c=IN IP4 " + localIp + "\r\n" +
                "t=0 0\r\n" +
                "m=audio " + rtpPort + " RTP/AVP " + PT_PCMU + " 101\r\n" +
                "a=rtpmap:" + PT_PCMU + " PCMU/8000\r\n" +
                "a=ptime:20\r\n" +
                "a=sendrecv\r\n" +
                "a=rtpmap:101 telephone-event/8000\r\n" +
                "a=fmtp:101 0-15";
    }

    private static String buildAmrOffer(String localIp, int rtpPort, int codecType) {
        int pt = PT_AMR_DYNAMIC;
        String codecName = codecType == CODEC_TYPE_WB ? "AMR-WB" : "AMR";
        int sampleRate = codecType == CODEC_TYPE_WB ? 16000 : 8000;

        return "v=0\r\n" +
                "o=- " + System.currentTimeMillis() + " 1 IN IP4 " + localIp + "\r\n" +
                "s=SipPhone\r\n" +
                "c=IN IP4 " + localIp + "\r\n" +
                "t=0 0\r\n" +
                "m=audio " + rtpPort + " RTP/AVP " + pt + " " + (pt + 1) + "\r\n" +
                "a=rtpmap:" + pt + " " + codecName + "/" + sampleRate + "\r\n" +
                "a=fmtp:" + pt + " octet-align=1; mode-change-capability=2; max-red=0\r\n" +
                "a=ptime:20\r\n" +
                "a=sendrecv\r\n" +
                "a=rtpmap:" + (pt + 1) + " telephone-event/8000\r\n" +
                "a=fmtp:" + (pt + 1) + " 0-15";
    }

    public static String buildAnswer(String localIp, int rtpPort, String remoteSdp, String preferredCodec) {
        SdpMediaInfo parsed = parseRemoteSdp(remoteSdp);
        if (parsed != null) {
            return buildOffer(localIp, rtpPort, parsed.codecName());
        }
        return buildOffer(localIp, rtpPort, preferredCodec);
    }

    public static SdpMediaInfo parseRemoteSdp(String sdp) {
        Matcher connMatch = Pattern.compile("c=IN IP4 (\\S+)").matcher(sdp);
        Matcher mediaMatch = Pattern.compile("m=audio (\\d+) ").matcher(sdp);
        if (!connMatch.find() || !mediaMatch.find()) return null;

        String remoteIp = connMatch.group(1);
        int remotePort = Integer.parseInt(mediaMatch.group(1));

        Matcher amrWb = Pattern.compile("a=rtpmap:(\\d+)\\s+AMR-WB/").matcher(sdp);
        Matcher amrNb = Pattern.compile("a=rtpmap:(\\d+)\\s+AMR/").matcher(sdp);
        Matcher pcmu = Pattern.compile("a=rtpmap:(\\d+)\\s+PCMU/").matcher(sdp);

        if (amrWb.find()) {
            return new SdpMediaInfo(remoteIp, remotePort, remotePort + 1,
                    Integer.parseInt(amrWb.group(1)), CODEC_TYPE_WB, CODEC_AMR_WB);
        } else if (amrNb.find()) {
            return new SdpMediaInfo(remoteIp, remotePort, remotePort + 1,
                    Integer.parseInt(amrNb.group(1)), CODEC_TYPE_NB, CODEC_AMR_NB);
        } else if (pcmu.find()) {
            return new SdpMediaInfo(remoteIp, remotePort, remotePort + 1,
                    Integer.parseInt(pcmu.group(1)), -1, CODEC_PCMU);
        } else {
            // Check for static PT 0 in m= line
            if (sdp.contains("m=audio " + remotePort + " RTP/AVP 0")) {
                return new SdpMediaInfo(remoteIp, remotePort, remotePort + 1, 0, -1, CODEC_PCMU);
            }
        }
        return null;
    }

    public record SdpMediaInfo(
            String remoteIp, int remoteRtpPort, int remoteRtcpPort,
            int payloadType, int codecType, String codecName) {}
}

File: app/src/main/java/com/telcobright/sipphone/verto/SdpBuilder.kt
package com.telcobright.sipphone.verto

import com.telcobright.sipphone.media.NativeMediaEngine

/**
 * Builds SDP offers/answers with AMR and PCMU codecs for Verto/FreeSWITCH.
 */
object SdpBuilder {

    const val CODEC_PCMU = "PCMU"
    const val CODEC_AMR_NB = "AMR-NB"
    const val CODEC_AMR_WB = "AMR-WB"

    /* PCMU is payload type 0 (static), AMR uses dynamic 96+ */
    private const val PT_PCMU = 0
    private const val PT_AMR_DYNAMIC = 96

    /**
     * Build an SDP offer for the selected codec.
     */
    fun buildOffer(
        localIp: String,
        rtpPort: Int,
        codec: String = CODEC_PCMU
    ): String {
        return when (codec) {
            CODEC_PCMU -> buildPcmuOffer(localIp, rtpPort)
            CODEC_AMR_NB -> buildAmrOffer(localIp, rtpPort, NativeMediaEngine.CODEC_TYPE_NB)
            CODEC_AMR_WB -> buildAmrOffer(localIp, rtpPort, NativeMediaEngine.CODEC_TYPE_WB)
            else -> buildPcmuOffer(localIp, rtpPort)
        }
    }

    private fun buildPcmuOffer(localIp: String, rtpPort: Int): String {
        return buildString {
            appendLine("v=0")
            appendLine("o=- ${System.currentTimeMillis()} 1 IN IP4 $localIp")
            appendLine("s=SipPhone")
            appendLine("c=IN IP4 $localIp")
            appendLine("t=0 0")
            appendLine("m=audio $rtpPort RTP/AVP $PT_PCMU 101")
            appendLine("a=rtpmap:$PT_PCMU PCMU/8000")
            appendLine("a=ptime:20")
            appendLine("a=sendrecv")
            appendLine("a=rtpmap:101 telephone-event/8000")
            appendLine("a=fmtp:101 0-15")
        }.trimEnd()
    }

    private fun buildAmrOffer(localIp: String, rtpPort: Int, codecType: Int): String {
        val pt = PT_AMR_DYNAMIC
        val codecName = if (codecType == NativeMediaEngine.CODEC_TYPE_WB) "AMR-WB" else "AMR"
        val sampleRate = if (codecType == NativeMediaEngine.CODEC_TYPE_WB) 16000 else 8000

        return buildString {
            appendLine("v=0")
            appendLine("o=- ${System.currentTimeMillis()} 1 IN IP4 $localIp")
            appendLine("s=SipPhone")
            appendLine("c=IN IP4 $localIp")
            appendLine("t=0 0")
            appendLine("m=audio $rtpPort RTP/AVP $pt ${pt + 1}")
            appendLine("a=rtpmap:$pt $codecName/$sampleRate")
            appendLine("a=fmtp:$pt octet-align=1; mode-change-capability=2; max-red=0")
            appendLine("a=ptime:20")
            appendLine("a=sendrecv")
            appendLine("a=rtpmap:${pt + 1} telephone-event/8000")
            appendLine("a=fmtp:${pt + 1} 0-15")
        }.trimEnd()
    }

    /**
     * Build SDP answer matching remote codec preference.
     */
    fun buildAnswer(
        localIp: String,
        rtpPort: Int,
        remoteSdp: String,
        preferredCodec: String = CODEC_PCMU
    ): String {
        val parsed = parseRemoteSdp(remoteSdp)
        return if (parsed != null) {
            buildOffer(localIp, rtpPort, parsed.codecName)
        } else {
            buildOffer(localIp, rtpPort, preferredCodec)
        }
    }

    /**
     * Parse remote SDP to extract media info.
     */
    fun parseRemoteSdp(sdp: String): SdpMediaInfo? {
        val connectionMatch = Regex("""c=IN IP4 (\S+)""").find(sdp) ?: return null
        val mediaMatch = Regex("""m=audio (\d+) """).find(sdp) ?: return null

        val pcmuMatch = Regex("""a=rtpmap:(\d+)\s+PCMU/""").find(sdp)
        val amrWbMatch = Regex("""a=rtpmap:(\d+)\s+AMR-WB/""").find(sdp)
        val amrNbMatch = Regex("""a=rtpmap:(\d+)\s+AMR/""").find(sdp)

        val (payloadType, codecType, codecName) = when {
            amrWbMatch != null -> Triple(
                amrWbMatch.groupValues[1].toInt(),
                NativeMediaEngine.CODEC_TYPE_WB,
                CODEC_AMR_WB
            )
            amrNbMatch != null -> Triple(
                amrNbMatch.groupValues[1].toInt(),
                NativeMediaEngine.CODEC_TYPE_NB,
                CODEC_AMR_NB
            )
            pcmuMatch != null -> Triple(
                pcmuMatch.groupValues[1].toInt(),
                -1, // PCMU handled differently
                CODEC_PCMU
            )
            else -> {
                /* Check if m= line contains 0 (PCMU static PT) */
                val mLine = mediaMatch.value
                if (mLine.contains(" 0 ") || mLine.endsWith(" 0")) {
                    Triple(0, -1, CODEC_PCMU)
                } else {
                    return null
                }
            }
        }

        return SdpMediaInfo(
            remoteIp = connectionMatch.groupValues[1],
            remoteRtpPort = mediaMatch.groupValues[1].toInt(),
            remoteRtcpPort = mediaMatch.groupValues[1].toInt() + 1,
            payloadType = payloadType,
            codecType = codecType,
            codecName = codecName
        )
    }

    data class SdpMediaInfo(
        val remoteIp: String,
        val remoteRtpPort: Int,
        val remoteRtcpPort: Int,
        val payloadType: Int,
        val codecType: Int,
        val codecName: String
    )
}
